File: src/main/golang/leetcode/editor/cn/[20]Valid Parentheses.go
package lc

//leetcode submit region begin(Prohibit modification and deletion)
func isValid(s string) bool {
	var stack []rune
	leftOf := func(c rune) rune {
		if c == ']' {
			return '['
		} else if c == '}' {
			return '{'
		} else {
			return '('
		}
	}
	for _, c := range s {
		if c == '[' || c == '(' || c == '{' {
			stack = append(stack, c)
		} else {
			if len(stack) > 0 && stack[len(stack)-1] == leftOf(c) {
				stack = stack[:len(stack)-1]
			} else {
				return false
			}
		}
	}
	return len(stack) == 0
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[20]Valid Parentheses.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def isValid(self, s: str) -> bool:
        stack = ['?']
        dic = {'(': ')', '[': ']', '{': '}', '?': '?'}
        for c in s:
            if c in dic:
                stack.append(c)
            elif dic[stack.pop()] != c:
                return False
        return len(stack) == 1

# leetcode submit region end(Prohibit modification and deletion)
